lib/collection/src/shards/local_shard.rs

```diff
@@ -395,9 +395,32 @@ impl LocalShard {
         bar.set_style(progress_style);
 
         bar.set_message(format!("Recovering collection {collection_id}"));
+
+        // To optimize loading we want to start reading (and applying) WAL from the last operation
+        // persisted by the storage.
+        //
+        // Each `Segment` is versioned individually, so there's no *single* storage version.
+        // We can guarantee that the *minimal* (i.e., "oldest") persisted version among *all*
+        // `Segment`s has been successfully applied to *all* of them.
+        //
+        // There's no public interface to access `Segment::persisted_version`, but
+        // `LocalShard::load` loads segments state from disk and then calls
+        // `LocalShard::load_from_wal`. So, as each `Segment` was *just* loaded from disk
+        // and was not modified in any way yet, `Segment::version` (returned by
+        // `<Segment as SegmentEntry>::version`) and `Segment::persisted_version` are the same.
+
         let segments = self.segments();
-        // ToDo: Start from minimal applied version
-        for (op_num, update) in wal.read_all() {
+
+        let minimal_persisted_version = segments
```

Comment:
Maybe it's better to do `.min().unwrap_or(0);`. That way we remove the option, and don't require the _ugly_ `Some(...)` in `.then_some(...)`.

Reply:
That's definitely true, but the result will be the same because we do `<= 0`.

Reply:
Isn't it backwards?

- imagine we don't have any segments, but have a bunch of OPs in WAL.
- so the `minimal_persisted_version` is `None`
- and the first OP in WAL is `0`
- if we do `None.unwrap_or(0) <= 0` we get `true`, and so the first entry in the WAL would be skipped

Though, you are right pointing my attention back to it, cause:
- `Segment::version` implementation does exactly that: `self.version.unwrap_or(0)`
- so there's a "collision": segment with version `0` may have _or may not_ have OP `0` applied

Reply:
You're totally right! I was too quick. :upside_down_face: